crux-llvm/src/Crux/LLVM/Simulate.hs

```diff
@@ -228,18 +239,64 @@ prepLLVMModule llvmOpts halloc sym bcFile memVar = do
 
 
 checkFun ::
+  forall arch msgs personality sym.
+  (IsSymInterface sym, ArchOk arch) =>
   Crux.Logs msgs =>
   Log.SupportsCruxLLVMLogMessage msgs =>
-  String -> ModuleCFGMap -> OverM personality sym LLVM ()
-checkFun nm mp =
+  LLVMOptions ->
+  ModuleTranslation arch ->
+  PreppedLLVM sym ->
+  OverM personality sym LLVM ()
+checkFun llvmOpts trans prepped =
   case Map.lookup (fromString nm) mp of
     Just (_, AnyCFG anyCfg) ->
       case cfgArgTypes anyCfg of
-        Empty ->
-          do liftIO $ Log.sayCruxLLVM (Log.SimulatingFunction (Text.pack nm))
-             (callCFG anyCfg emptyRegMap) >> return ()
-        _     -> throwCError BadFun  -- TODO(lb): Suggest uc-crux-llvm?
+        Empty -> simulateFun anyCfg emptyRegMap
+
+        (Empty :> LLVMPointerRepr w :> PtrRepr)
+          |  isMain, supplyMainArguments llvmOpts
+          ,  Just Refl <- testEquality w (knownNat @32)
+          -> checkMainWithArguments anyCfg
+
+        _ -> throwCError (BadFun nm isMain)  -- TODO(lb): Suggest uc-crux-llvm?
     Nothing -> throwCError (MissingFun nm)
+  where
+    nm     = entryPoint llvmOpts
+    mp     = cfgMap trans
+    memVar = prepMemVar prepped
+    mem    = prepMem prepped
+    isMain = nm == "main"
+
+    simulateFun :: CFG LLVM blocks ctx ret ->
+                   RegMap sym ctx ->
+                   OverM personality sym LLVM ()
+    simulateFun anyCfg args = do
+      liftIO $ Log.sayCruxLLVM (Log.SimulatingFunction (Text.pack nm))
+      callCFG anyCfg args >> return ()
+
+    -- Simulate a function with the signature `int main(int argc, char* argv[])`,
+    -- supplying argc=0 and argv={} arguments.
+    checkMainWithArguments ::
+      CFG LLVM blocks (EmptyCtx ::> LLVMPointerType 32 ::> TPtr arch) ret ->
+      OverM personality sym LLVM ()
+    checkMainWithArguments anyCfg = do
+      ctx <- getContext
+      let w         = knownNat @32
+          sym       = ctx^.ctxSymInterface
+          dl        = llvmDataLayout (trans^.transContext.llvmTypeCtx)
+          tp        = ArrayType 0 (PtrType (MemType i8))
+          tp_sz     = memTypeSize  dl tp
+          alignment = memTypeAlign dl tp
+          loc       = "crux-llvm main(argc, argv) simulation"
+
+      argc <- liftIO $ LLVMPointer <$> natLit sym 0 <*> bvLit sym w (BV.zero w)
+      sz   <- liftIO $ bvLit sym PtrWidth $ bytesToBV PtrWidth tp_sz
+      (argv, mem') <- liftIO $ doAlloca sym mem sz alignment loc
```

Comment:
Could you comment on why `doAlloca` over just `doCalloc`?

Reply:
My goal was to simulate a function that looks roughly like this:

```hs
int crucible_main() {
  int argc = 0;
  char *argv[] = {};
  return main(argc, argv);
}
```

With this in mind, `doAlloca` seemed like a more natural fit, since the LLVM bitcode to create `argv` will use the `alloca` instruction anyway. There's no reason I couldn't use `doCalloc`, I suppose, but then the memory would be on the heap instead of on the stack. That could potentially have ramifications if we ever enter the [`valid-memcleanup`/`valid-memsafety`](https://github.com/GaloisInc/crucible/issues/790) categories (which care a lot about the lifetime of heap-allocated data), so `doAlloca` seemed like the simpler approach.

Having typed all of this out, I should turn this into a comment in the source code as well.

Reply:
I think that is reasonable. `argv` *does* actually live on the stack in most binaries (on Linux, in the aux vector), though in a different spot, so it is reasonable. I was just curious what level of fidelity you were looking for.

Reply:
I've turned https://github.com/GaloisInc/crucible/pull/826/files#r698825652 into a comment above `checkMainWithArguments`.